src/erlcloud_aws.erl

```diff
@@ -289,19 +289,81 @@ format_timestamp({{Yr, Mo, Da}, {H, M, S}}) ->
 %% available in the environment, if available.  If no credentials are
 %% available then this will just return a default <code>#aws_config{}</code>
 %% record.
-%%
+%% The credentials are; Id, Key, Token, and Region. For each credential we
+%% will first look for the environment variables "AWS_ACCESS_KEY_ID",
+%% "AWS_SECRET_ACCESS_KEY", "AWS_SECURITY_TOKEN", and "AWS_REGION",
+%% respectively. If no value is found, we look at the erlcloud application
+%% environment for the keys aws_access_key_id, aws_secret_access_key,
+%% aws_security_token, and aws_region, respectively. If still no value is
+%% found, it will default to 'undefined'.
+%% Id, Key, and Token will be set in #aws_config{}, in the fields
+%% access_key_id, secret_access_key, and security_token, respectively.
+%% If Region is set, we will attempt to set the appropriate URL for this
+%% region for each service.
+
+%% check the cache
 default_config() ->
     case get(aws_config) of
-        undefined -> default_config_env();
+        undefined -> default_config_wrap();
         Config -> Config
     end.
 
-default_config_env() ->
-    case config_env() of
-        {ok, Config} -> Config;
-        {error, _} -> #aws_config{}
+default_config_wrap() ->
+    {Id, Key, Token, Region} = default_config_get(),
+    default_config_region(default_config_assert(Id, Key, Token), Region).
+
+default_config_get() ->
+    {default_config_get("AWS_ACCESS_KEY_ID", aws_access_key_id),
+     default_config_get("AWS_SECRET_ACCESS_KEY", aws_secret_access_key),
+     default_config_get("AWS_SECURITY_TOKEN", aws_security_token),
+     default_config_get("AWS_REGION", aws_region)}.
+
+%% try to get config value from OS env, failing that try app env, else
+%% return undefined
+default_config_get(OSENV, APPENV) ->
+    case default_config_osenv(OSENV) of
+        undefined -> default_config_appenv(APPENV);
+        Val -> Val
+    end.
+
+default_config_osenv(ENV) ->
+    case os:getenv(ENV) of
+        Value when is_list(Value) -> Value;
+        false -> undefined
+    end.
+
+default_config_appenv(ENV) ->
+    case application:get_env(erlcloud, ENV) of
+        {ok, Value} -> Value;
+        undefined -> undefined
+    end.
+
+default_config_assert(Id, Key, Token) ->
```

Comment:
your code does not have the same semantics as the original (which required that neither Id nor Key was undefined).
Is that a change you want to make?